### ISSUE TITLE
[portal] Add captcha to login page

### ISSUE THREAD

**Issue by yeganx** [gravitee-io/issues#1787]
hi,
I want add captch to login component.

I can add it to index page with success with this module
https://github.com/VividCortex/angular-recaptcha
but can not add to login component.

**Comment by NicolasGeraud** [comment]
hi @yeganx ,

why can't you add it to login ?
Please also add a settings to enable this new feature (must be disable by default)

**Comment by yeganx** [comment]
I try add capcha throught https://github.com/VividCortex/angular-recaptcha
here is my code for index :
  
<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <title ng-bind="portalTitle"></title>
  <meta name="description" content="">
  <meta name="viewport" content="width=device-width">
  <link rel="icon" href="favicon.ico" />
  <style type="text/css">
    [ng\:cloak],
    [ng-cloak],
    [data-ng-cloak],
    [x-ng-cloak],
    .ng-cloak,
    .x-ng-cloak,
    .ng-hide:not(.ng-hide-animate) {
      display: none !important;
    }

    .gravitee-splash-screen {
      background-color: rgb(45, 50, 62);
      bottom: 0;
      left: 0;
      position: absolute;
      right: 0;
      top: 0;
      z-index: 99999;
    }

    img.gravitee-splash-screen {
      color: silver;
      display: flex;
      margin: 0 auto;
      max-width: 200px;
      position: relative;
      top: 40%;
    }

    .animate-show-hide.ng-hide {
      opacity: 0;
    }

    .animate-show-hide.ng-hide-add,
    .animate-show-hide.ng-hide-remove {
      transition: all linear 0.5s;
    }

    #gravitee-error {
      color: red;
      font-size: 20px;
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
      z-index: 999999;
    }
  </style>

  <script language="JavaScript">
    /**
     * detect IE
     * https://codepen.io/gapcode/pen/vEJNZN
     *
     * returns version of IE or false, if browser is not Internet Explorer
     */
    function detectIE() {
      let ua = window.navigator.userAgent;
      let msie = ua.indexOf('MSIE ');
      if (msie > 0) {
        // IE 10 or older => return version number
        return parseInt(ua.substring(msie + 5, ua.indexOf('.', msie)), 10);
      }
      let trident = ua.indexOf('Trident/');
      if (trident > 0) {
        // IE 11 => return version number
        let rv = ua.indexOf('rv:');
        return parseInt(ua.substring(rv + 3, ua.indexOf('.', rv)), 10);
      }
      let edge = ua.indexOf('Edge/');
      if (edge > 0) {
        // Edge (IE 12+) => return version number
        return parseInt(ua.substring(edge + 5, ua.indexOf('.', edge)), 10);
      }
      // other browser
      return false;
    }

    window.onload = function () {
      let isIE = detectIE();
      if (isIE !== false && isIE < 12) {
        document.getElementById("BrowserNotSupportedBanner").style.display = "";
      }
    }
  </script>
  <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.7.2/angular.min.js"></script>
  <script src="./angular-no-captcha.js"></script>
  <script>
    var SITE_KEY = '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI';

    var app = angular
      .module('myApp', ['noCAPTCHA'])
      .config(['noCAPTCHAProvider',
        function (noCaptchaProvider) {
          noCaptchaProvider.setSiteKey(SITE_KEY);
          noCaptchaProvider.setTheme('light');
        }
      ])
      .controller('exampleController', [
        '$scope',
        function exampleController($scope) {
          $scope.gRecaptchaResponse = '';

          $scope.$watch('gRecaptchaResponse', function () {
            $scope.expired = false;
          });

          $scope.expiredCallback = function expiredCallback() {
            $scope.expired = true;
          };
        }]);
  </script>
</head>

<body layout="column">
  <p id="BrowserNotSupportedBanner" class="browsehappy" style="display: none">You are using an <strong>outdated</strong>
    browser. Please <a href="http://browsehappy.com/">upgrade your browser</a> to improve your experience.</p>

  <div id="gravitee-error"></div>

  <div id="loader" ng-show="displayLoader" class="gravitee-splash-screen animate-show-hide"></div>

  <gravitee-navbar></gravitee-navbar>
  <div flex ui-view></div>
  <div ng-app="myApp">
    <no-captcha expired-callback="expiredCallback" g-recaptcha-response="gRecaptchaResponse" theme='light'></no-captcha>

    <code ng-class="{ 'expired': expired }">
      {{ gRecaptchaResponse }}
    </code>
  </div>
</body>

</html>
that is work fine ,
but login companent dont work properly.
login code : 
<md-content class="gravitee-no-sidenav-container" flex>
  <div class="login-page">
    <div class="form">

      <div class="gravitee-no-sidenav-header" layout="column">
        <label>Login</label>
      </div>

      <form name="formUser" ng-submit="$ctrl.login()" layout="column" ng-if="!$ctrl.localLoginDisabled">

        <div layout="row">
          <ng-md-icon icon="person" style="fill: #555"></ng-md-icon>
          <md-input-container class="md-icon-float md-block" flex>
            <label>Username</label>
            <input ng-model="$ctrl.user.username" type="text" required autofocus>
          </md-input-container>
        </div>

        <div layout="row">
          <ng-md-icon icon="vpn_key" style="fill: #555"></ng-md-icon>
          <md-input-container class="md-icon-float md-block" flex>
            <label>Password</label>
            <input ng-model="$ctrl.user.password" type="password" required>
          </md-input-container>
        </div>

        <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.7.2/angular.min.js"></script>
        <script src="../../angular-no-captcha.js"></script>
        <script>
          var SITE_KEY = '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI';

          var app = angular
            .module('myApp', ['noCAPTCHA'])
            .config(['noCAPTCHAProvider',
              function (noCaptchaProvider) {
                noCaptchaProvider.setSiteKey(SITE_KEY);
                noCaptchaProvider.setTheme('light');
              }
            ])
            .controller('exampleController', [
              '$scope',
              function exampleController($scope) {
                $scope.gRecaptchaResponse = '';

                $scope.$watch('gRecaptchaResponse', function () {
                  $scope.expired = false;
                });

                $scope.expiredCallback = function expiredCallback() {
                  $scope.expired = true;
                };
              }]);
        </script>
        <div ng-app="myApp" ng-controller="exampleController">
          <no-captcha expired-callback="expiredCallback" g-recaptcha-response="gRecaptchaResponse" theme='light'></no-captcha>
          <code ng-class="{ 'expired': expired }">
              {{ gRecaptchaResponse }}
          </code>
        </div>

        <md-button type="submit" class="md-raised md-primary" ng-disabled="formUser.$invalid || formUser.$pristine">Log
          in</md-button>


        <br>
        <p class="text-center text-muted" ng-if="$ctrl.userCreationEnabled">
          <small>Don't have an account yet? <a ui-sref="registration">Sign up</a></small>
        </p>

        <div ng-if="$ctrl.providers.length > 0">
          <div class="signup-or-separator">
            <h6 class="text">or</h6>
            <hr>
          </div>
        </div>
      </form>

      <button class="btn btn-block btn-{{provider.icon}}" style="background-color: {{provider.color}}; color: #fff;"
        data-ng-repeat="provider in $ctrl.identityProviders" ng-click="$ctrl.authenticate(provider.id)">
        <ng-md-icon icon="{{provider.icon}}" size="20" style="fill: white;margin-right: 5px"></ng-md-icon>Sign in with
        {{provider.name}}
      </button>
    </div>
  </div>
</md-content>

here is the result :
https://pasteboard.co/HVtvZLD.jpg

**Comment by brasseld** [comment]
Hi @yeganx 

May I ask you to create a branch on your own fork of the management-UI and which is including these changes?

It will be easier for us to test it.

**Comment by brasseld** [comment]
Hi @yeganx 

What are the news here ?

**Comment by yeganx** [comment]
it is the fork url : https://github.com/yeganx/gravitee-management-webui in test branch
changes are in gravitee-management-webui/src/index.html  
and gravitee-management-webui/src/user/login/login.html

**Comment by yeganx** [comment]
can any body fix it ?

**Comment by brasseld** [comment]
No one took a look for now.

There is more than one issue to manage here...

**Comment by yeganx** [comment]
is there any support team for users that can answer online ?

**Comment by brasseld** [comment]
https://gitter.im/gravitee-io/gravitee-io

**Comment by aelamrani** [comment]
Why not using an angularjs module like:
https://github.com/VividCortex/angular-recaptcha 

This way, you will be able to integrate such feature easily. Wdyt?

**Comment by brasseld** [comment]
@aelamrani it was the purpose of this comment: https://github.com/gravitee-io/issues/issues/1787#issuecomment-452241336

So @yeganx has pushed a fork / branch with this module but this later is not working as expected.

**Comment by brasseld** [comment]
Hi @yeganx 

You're not using the recaptcha module correctly.
That's the reason why you're not able to login anymore.

**Comment by aelamrani** [comment]
Ok i missed it because i was focused on the snippet below and it seems like the module is not well used.

**Comment by yeganx** [comment]
> 
> 
> Hi @yeganx
> 
> You're not using the recaptcha module correctly.
> That's the reason why you're not able to login anymore.

hi @brasseld  I dont know how should I use it ?
none of my script worked in login page !  but same script work in index page .

**Comment by brasseld** [comment]
If you have a look to your dev console from your browser, you will see an error saying that your captchaController does not exist. This is the root cause of your issue.

**Comment by nemati21** [comment]
@ @Hello. 
I defined a controller in a script tag and assign it to "ng-controller of captcha tag at login component , but it does'' not work and even change style of     @@@@button'.
I' know angularjs very little.
Please help me how to do it.
Thanks.

**Comment by nemati21** [comment]
@ @@@Hello. 
I defined a controller in a script tag and assign it to "ng-controller of captcha tag at login component , but it does'' not work and even change style of button'.
I' know angularjs very little.
Please help me how to do  @it.
Thanks.

**Comment by brasseld** [comment]
I suppose you have to make some changes in login component / controller ?

**Comment by nemati21** [comment]
Yes, I want to create a google captcha  in login component but I do not know how to define controller in login component. At all of examples, this controller is defined in script tag. This works in index page but  does not work in login component.

**Comment by brasseld** [comment]
There is already a `LoginController` available so why would you want to create a new one ?

**Comment by nemati21** [comment]
@google captcha for some settings needs a controller and these setting can be defined in script tag. 
Any script tag does not work in login component even a console.log().
This is my problem.

**Comment by nemati21** [comment]
I can see google captcha tag (rCaptcha or no-captcha) at login component (in dev console) but it is empty because it can not connect to google API. 
This problem is just component and I can create google captcha in index.html.

**Comment by nemati21** [comment]
I a'm so sorry for the delay in answering.
If I want to explain my problem with details and more information, I need to add a google captcha (for example "I am not robot) in login page/component. 
I can add it in index page that show every page but it does not sho in login component (top of login button).
I am beginner at Angular .
Thanks
